File: hellospring/src/test/java/kopher/hellospring/payment/PaymentServiceTest.java
package kopher.hellospring.payment;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.math.BigDecimal;
import java.time.Clock;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;

import static org.assertj.core.api.Assertions.assertThat;

class PaymentServiceTest {
    Clock clock;


    @BeforeEach
    void setUp() {
        this.clock = Clock.fixed(Instant.now(), ZoneId.systemDefault());
    }

    @Test
    @DisplayName("prepare 메소드가 요구사항 3가지를 잘 충족했는지 검증")
    void convertedAmount() throws IOException {
        Clock clock = Clock.fixed(Instant.now(), ZoneId.systemDefault());

        testAmount(BigDecimal.valueOf(500), BigDecimal.valueOf(5_000), clock);
        testAmount(BigDecimal.valueOf(1_000), BigDecimal.valueOf(10_000), clock);
        testAmount(BigDecimal.valueOf(3_000), BigDecimal.valueOf(30_000), clock);
    }

    @Test
    void validUntil() throws IOException {
        PaymentService paymentService = new PaymentService(new ExRateProviderStub(BigDecimal.valueOf(1_000)), clock);

        Payment payment = paymentService.prepare(1L, "USD", BigDecimal.TEN);

        assertThat(payment.getValidUntil()).isEqualTo(LocalDateTime.now(clock).plusMinutes(30));
    }

    private static void testAmount(BigDecimal exRate, BigDecimal convertedAmount, Clock clock) throws IOException {
        // 제어 가능한 ExRateProviderStub 테스트 대역을 사용
        PaymentService paymentService = new PaymentService(new ExRateProviderStub(exRate), clock);

        Payment payment = paymentService.prepare(1L, "USD", BigDecimal.TEN);


        assertThat(payment.getExRate()).isEqualByComparingTo(exRate);
        assertThat(payment.getConvertedAmount()).isEqualTo(convertedAmount);
    }

}

File: spring-principle/src/test/kotlin/jp/kopher/springprinciple/payment/PaymentServiceTest.kt
package jp.kopher.springprinciple.payment

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.math.BigDecimal
import java.time.Clock
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId

class PaymentServiceTest {

    private lateinit var clock: Clock

    @BeforeEach
    fun beforeEach() {
        clock = Clock.fixed(Instant.now(), ZoneId.systemDefault())
    }

    @Test
    fun `요구사항 3가지를 잘 충족했는지 검증`(){

        testAmount(exRate = 500.toBigDecimal(), convertedAmount = 5_000.toBigDecimal(), clock = clock)
    }

    @Test
    fun validUntil() {
        val paymentService = PaymentService(ExRateProviderStub(1_000.toBigDecimal()), clock)
        val payment = paymentService.prepare(1L, "USD", BigDecimal.TEN)

        val now = LocalDateTime.now(clock)
        val expectedValidUntil = now.plusMinutes(30)

        assertThat(payment.validUntil).isAfter(expectedValidUntil)
    }

    private fun testAmount(exRate: BigDecimal, convertedAmount: BigDecimal, clock: Clock) {
        val paymentService = PaymentService(ExRateProviderStub(exRate), clock)

        val payment = paymentService.prepare(1L, "USD", BigDecimal.TEN)

        // BigDecimal 는 isEqualByComparingTo 를 사용해야 한다.
        assertThat(payment.exRate).isEqualByComparingTo(exRate)
        assertThat(payment.convertedAmount).isEqualByComparingTo(convertedAmount)
    }

}